### PR TITLE
Support changing the ROS_DOMAIN_ID in OpenSplice without changing the config file

### PR DESCRIPTION
**Pull request by wjwwood**
Currently you have to update the system installed OpenSplice config file, or override it using `OSPL_URI` in order to use a different `ROS_DOMAIN_ID`. This is out of sync with the other implementations and inconvenient.

So this pull request works around this issue by copying the default configuration for OpenSplice into the `opensplice_cmake_module` package, replacing the default domain id (`0`) with `${ROS_DOMAIN_ID}`, installing it to the workspace, and changing the environment hooks to default to referencing this file rather than the built-in one. This means that when this config file is evaluated by OpenSplice it will substitute the `${ROS_DOMAIN_ID}` in the config file with what ever is in the environment. This strategy is based on the feedback I got on their forum:

http://forums.opensplice.org/index.php?/topic/2591-programmatically-choosing-the-domain-id/#entry4471

The other thing we have to do is to ensure that the `ROS_DOMAIN_ID` is set before trying to use OpenSplice, and therefore forcing the configuration file to be evaluated. See https://github.com/ros2/rclcpp/pull/102

Still testing, not quite ready for review.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
Instead of requiring `rclcpp` to handle this case `opensplice_cmake_module` could simply provide an environment hook which ensures that the domain id is set for OpenSplice.

**Comment by wjwwood** [comment]
@dirk-thomas Someone could also unset `ROS_DOMAIN_ID`. In which case `connext` would work, defaulting to `0` and OpenSplice would not.

**Comment by dirk-thomas** [comment]
Someone could also unset `OSPL_URI` after sourcing the environment and that would also break. I guess there is always a way to break the system if you try :wink:

**Comment by wjwwood** [comment]
That's OpenSplice specific, the `ROS_DOMAIN_ID` is ROS specific and having it set or not set should result in the same behavior, regardless of the implementation being used. So I think the argument that you can break the system if you try is not an argument which applies to this pr.

As to where we ensure the `ROS_DOMAIN_ID` is set for OpenSplice, I put it in `rclcpp` because we read the `ROS_DOMAIN_ID` in `rclcpp`. I can look at moving it to the `rmw_opensplice` implementation of create_node.

**Comment by wjwwood** [comment]
I moved it to `rmw_opensplice_cpp`, the benefit being that I can scope the use of this more tightly. So now if the `ROS_DOMAIN_ID` is not set, it sets it, creates the participant factory, and then unsets it again. If the ROS_DOMAIN_ID is set, but different from the `domain_id` that was passed in (something we might want to expose through `rclcpp` in the future), then it sets it to what was passed in, storing the current value, and restores it after it finishes creating the domain participant factory.

I still need to test this locally.

**Comment by wjwwood** [comment]
I also added a check to see if the `OSPL_URI` is set, and returns a relevant error message if it is not. Previously if `OSPL_URI` was not set you just get "failed to create participant".

**Comment by wjwwood** [comment]
I've tested it locally:

```
william@farl ~
% bash
william@farl:~
$ source ./ros2/install/setup.bash
william@farl:~
$ talker__rmw_opensplice_cpp
Publishing: 'Hello World: 1'
Publishing: 'Hello World: 2'
Publishing: 'Hello World: 3'
Publishing: 'Hello World: 4'
Publishing: 'Hello World: 5'
^Csignal_handler(2)
william@farl:~
$ ROS_DOMAIN_ID=1 talker__rmw_opensplice_cpp
Publishing: 'Hello World: 1'
Publishing: 'Hello World: 2'
Publishing: 'Hello World: 3'
Publishing: 'Hello World: 4'
Publishing: 'Hello World: 5'
^Csignal_handler(2)
william@farl:~
$ exit
```

```
william@farl ~
% bash
william@farl:~
$ source ros2/install/setup.bash
william@farl:~
$ listener__rmw_opensplice_cpp
I heard: [Hello World: 1]
I heard: [Hello World: 2]
I heard: [Hello World: 3]
I heard: [Hello World: 4]
I heard: [Hello World: 5]
^Csignal_handler(2)
william@farl:~
$ ROS_DOMAIN_ID=2 listener__rmw_opensplice_cpp
^Csignal_handler(2)
william@farl:~
$ exit
```

I still need to test it manually on Windows, but the build is passing on Windows: http://ci.ros2.org/job/ros2_batch_ci_windows/377/

It has test failures, but I'll have to dig into them to know if that's expected or not.

File: opensplice_cmake_module/env_hook/opensplice.bat.in [hunk 1]
```diff
@@ -2,6 +2,7 @@ set "OSPL_HOME=@OpenSplice_HOME@"
 
 call "%OSPL_HOME%\release.bat" 1> nul
 
+set "OSPL_URI=file://@CMAKE_INSTALL_PREFIX@/share/opensplice_cmake_module/config/ros_ospl.xml"
```
Comment (dirk-thomas) [inline opensplice_cmake_module/env_hook/opensplice.bat.in]: Please don't use `CMAKE_INSTALL_PREFIX` since it break ability to relocate the install space. Use `$AMENT_CURRENT_PREFIX` instead.

File: rmw_opensplice_cpp/src/functions.cpp [hunk 2]
```diff
@@ -287,6 +291,52 @@ rmw_create_node(const char * name, size_t domain_id)
...
+#ifndef _WIN32
+  ret = setenv(env_var, domain_id_as_string.c_str(), 1);
+#else
+  ret = (SetEnvironmentVariable(env_var, domain_id_as_string.c_str()) != 0) ? 0 : 1;
+#endif
+  if (0 != ret) {
+    RMW_SET_ERROR_MSG("failed to set the ROS_DOMAIN_ID");
+    return nullptr;
+#ifdef _WIN32
+    free(ros_domain_id);
```
Comment (dirk-thomas) [inline rmw_opensplice_cpp/src/functions.cpp]: This line is unreachable.

File: opensplice_cmake_module/env_hook/opensplice.sh.in [hunk 3]
```diff
@@ -1,16 +1,29 @@
...
+  export OSPL_URI="$_DEFAULT_OSPL_URI"
+elif [ "$OSPL_URI" != "$_DEFAULT_OSPL_URI" ]; then
+  echo "Warning: OSPL_URI was already set to [[$OSPL_URI]]. This will not override it to the default [[$_DEFAULT_OSPL_URI]]. Please make sure this is the config that you want." 1>&2
+fi
+
+unset _DEFAULT_OSPL_URI
+
+# If release.com was not sourced...
+if [ -z "$OSPL_TMPL_PATH" ]; then
+  export OSPL_TMPL_PATH=$OSPL_HOME/etc/opensplice/idlpp
```
Comment (dirk-thomas) [inline opensplice_cmake_module/env_hook/opensplice.sh.in]: Why is this necessary now?

Comment (wjwwood) [inline opensplice_cmake_module/env_hook/opensplice.sh.in]: This isn't new, I just restructured the logic so it was moved further down. I'm not sure why it's needed, but it was already in there.

Comment (dirk-thomas) [inline opensplice_cmake_module/env_hook/opensplice.sh.in]: Can we then keep this line (which states is only done when release.com was not sourced) in the else block?

File: rmw_opensplice_cpp/src/functions.cpp (hunk 2 above)
Comment (wjwwood) [inline rmw_opensplice_cpp/src/functions.cpp]: Good catch, 162b53b

File: opensplice_cmake_module/env_hook/opensplice.bat.in (hunk 1 above)
Comment (wjwwood) [inline opensplice_cmake_module/env_hook/opensplice.bat.in]: Ok, ad456f8

**Comment by wjwwood** [comment]
This works on Linux, but is still not working correctly on Windows, I'll look into it today and make a decision about its feasibility as a fix on Windows.

File: opensplice_cmake_module/env_hook/opensplice.sh.in (hunk 3 above)
Comment (wjwwood) [inline opensplice_cmake_module/env_hook/opensplice.sh.in]: You mean this line and the rest of the stuff that comes after it? I can try to reorganize it again.

Comment (dirk-thomas) [inline opensplice_cmake_module/env_hook/opensplice.sh.in]: Line 26 was before done in line 6 as part of an else block. I think it should stay the same - in an else block of line 5 which would be inserted at line 14.

Comment (wjwwood) [inline opensplice_cmake_module/env_hook/opensplice.sh.in]: What about the lines that follow?

Before there was just if and then else. Now there is if then else, then something that has to be done in either case and then the rest (which used to be in the else case). I'll have to see if I can reorganize it to be in the origin else case.

Comment (wjwwood) [inline opensplice_cmake_module/env_hook/opensplice.sh.in]: I tried to simply the logic in c8bdad5. I'm testing it out now.

**Comment by wjwwood** [comment]
I've given up on getting it to work around on Windows, as I do on the other platforms. So now the Windows env hook sets the `ROS_DOMAIN_ID` is not set to `0` and errors out if it is not set when you run the program.

I'll do uncrustify and squash after lunch, CI for now:

- http://ci.ros2.org/job/ros2_batch_ci_linux/367/
- http://ci.ros2.org/job/ros2_batch_ci_osx/274/
- http://ci.ros2.org/job/ros2_batch_ci_windows/410/

**Comment by wjwwood** [comment]
CI looks good. The test failures on Windows seem to be unrelated. I just finished manually running the examples on Windows as well, setting, changing, and unsetting the `ROS_DOMAIN_ID` in turn.

I think this is ready for review. I'll squash after review.

**Comment by dirk-thomas** [comment]
LGTM

**Comment by dirk-thomas** [comment]
I "fixed" the indentation level of the xml file from 3 to 4 spaces since that is the format osplconf uses when writing files: 44278bd9955f1a40fef0acac7db68608fcd2251b
